src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3813,23 +3813,48 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
 
             void processResource(IOperation resource, ArrayBuilder<(IVariableDeclarationOperation, IVariableDeclaratorOperation)>? resourceQueueOpt)
             {
-                // When ResourceType is a non-nullable value type, the expansion is:
+                // When ResourceType is a non-nullable value type that implements IDisposable, the expansion is:
                 //
                 // {
                 //   ResourceType resource = expr;
                 //   try { statement; }
                 //   finally { ((IDisposable)resource).Dispose(); }
                 // }
                 //
-                // Otherwise, when Resource type is a nullable value type or
-                // a reference type other than dynamic, the expansion is:
+                // Otherwise, when ResourceType is a non-nullable value type that implements
+                // disposal via pattern, the expansion is:
+                //
+                // {
+                //   try { statement; }
+                //   finally { resource.Dispose(); }
+                // }
+                // Otherwise, when Resource type is a reference type other than dynamic
+                // that implements IDisposable, the expansion is:
                 //
                 // {
                 //   ResourceType resource = expr;
                 //   try { statement; }
                 //   finally { if (resource != null) ((IDisposable)resource).Dispose(); }
                 // }
                 //
+                // Otherwise, when Resource type is a reference type other than dynamic
+                // that implements disposal via pattern, the expansion is:
+                //
+                // {
+                //   ResourceType resource = expr;
+                //   try { statement; }
+                //   finally { if (resource != null) resource.Dispose(); }
+                // }
+                //
+                // Otherwise, when Resource type is a nullable value type
+                // that implements IDisposable, the expansion is:
+                //
+                // {
+                //   ResourceType resource = expr;
+                //   try { statement; }
+                //   finally { if (resource.HasValue()) ((IDisposable)resource.GetValueOrDefault()).Dispose(); }
```

Comment:
>finally { if (resource.HasValue()) ((IDisposable)resource.GetValueOrDefault()).Dispose(); } [](start = 21, length = 91)

Where do you see this in flow graph tests?

Reply:
Mentioned above. We don't the test shows something different: do we want to make the CFG represent the actual lowered value or what the spec says?

Reply:
> Mentioned above, we don't. The test shows something different: do we want to make the CFG represent the actual lowered value or what the spec says?

Does the comment accurately reflect CFG builder's behavior? If so, we don't want to make any changes for the comment. We also do not want to make any changes for scenarios that are not about pattern-based disposal, which is supposed to be the focus of this PR.

---
In reply to: [534616619](https://github.com/dotnet/roslyn/pull/49481#discussion_r534616619) [](ancestors = 534616619)

Reply:
Yes apologies. I was confused by your comment, I thought you were saying that the comment, and thus the tests are incorrect for those scenarios.